blogposts/2021/how-to-deconstruct-the-monolith.md

```diff
@@ -0,0 +1,106 @@
+
+## Step 3: Say goodbye to your monolith
+
+Once we split high traffic and high load parts of a monolith into microservices, we should split the rest of our monolith along functional lines. For example, this includes authentication, notification, and logging. To help with this task, let’s look at two beneficial microservices patterns: the [strangler pattern](https://martinfowler.com/bliki/StranglerFigApplication.html) and the [branch by abstraction pattern](https://martinfowler.com/bliki/BranchByAbstraction.html).
+
+### The strangler pattern 
+
+The strangler pattern allows us to migrate from a monolithic application to microservices incrementally. It does this by extracting parts of a monolith functionality and introducing them in the new service.
+
+Once we complete this process, the old monolith functionality is “strangled.” That is, we have made it obsolete and phased it out. We can sum this up in three steps:
+
+1. Identifying requests that we can intercept and creating a proxy to pass those requests to the original endpoint and back to the monolith without breaking the functionality
+2. Developing and migrating the monolith functionality to the new service
+3. Configuring the proxy to route the requests to the new service instead of the old endpoint exposed by the monolith
+
+As the strangler application grows and we replace more functionality with new services, the monolith shrinks over time.
+
+As you can imagine, the strangler pattern is the recommended approach for functionalities already exposed as API endpoints by the monolith. For example, this includes REST and GraphQL APIs and SOAP web services. The benefit of this pattern comes from the fact that we can migrate functionalities without even touching the monolith’s code. But, the feasibility of the strangler pattern depends on the level of modularity presented by the monolith. It also depends on whether or not it exposes its functionalities with public endpoints.
+
+### The branch by abstraction pattern
+
+When we can’t easily intercept calls and create a proxy to reroute the requests, we can benefit from the branch by abstraction technique. It’s like a strategy design pattern applied to a more extensive scope.
+
+Branch by abstraction requires us to modify the existing monolith’s code. It works by implementing the next few steps:
+
+1. Identify the code within the monolith that composes the functionality. This means finding out classes, methods, and parameters without worrying about the actual implementation. Finding those functional lines in a vast monolith codebase is often an arduous task. To make the lives of your microservices developers easier, consider [universal code search](https://about.sourcegraph.com/). It provides a search engine for code across all your repositories and branches, so you can find and filter algorithms, definitions, and references instantaneously.
```

Comment:
I don't think we have something better to link to here yet, but maybe we can replace it with the code search guide when it's live? I don't love sending people to our marketing site from here.

Reply:
@rebeccadee I agree. I searched for a blog post that might do the trick but couldn't find anything that was current that would work. Do you think it's too pushy to put in our homepage link? What about https://about.sourcegraph.com/get-started?

Reply:
```suggestion
1. Identify the code within the monolith that composes the functionality. This means finding out classes, methods, and parameters without worrying about the actual implementation. Finding those functional lines in a vast monolith codebase is often an arduous task. To make the lives of your microservices developers easier, consider [universal code search](/get-started). It provides a search engine for code across all your repositories and branches, so you can find and filter algorithms, definitions, and references instantaneously.
```